lib/ClangImporter/ImportDecl.cpp

```diff
@@ -1974,6 +1975,59 @@ applyPropertyOwnership(VarDecl *prop,
   }
 }
 
+/// Turn T into Unmanaged<T>, or into Unmanaged<T>? if `isOptional` is true.
+static Type getUnmanagedType(swift::ASTContext& Ctx,
+                             Type payloadType, bool isOptional) {
+  NominalTypeDecl *unmanagedDecl = Ctx.getUnmanagedDecl();
+  if (!unmanagedDecl || unmanagedDecl->getGenericParams()->size() != 1)
+    return payloadType;
+
+  Type result = BoundGenericType::get(unmanagedDecl, /*parent*/ Type(),
+                                      payloadType);
+  if (isOptional)
+    result = OptionalType::get(result);
+  return result;
+}
+
+static void
+applyPropertyOwnership(VarDecl *prop,
+                       clang::Qualifiers::ObjCLifetime lifetime) {
+  Type ty = prop->getInterfaceType();
```

Comment:
Do you not want to look though a ReferenceStorageType here? Because you may end up creating a nested weak type later down.

Reply:
I thought C records can have ARC references of various kinds which are imported as ReferenceStorageType.

Reply:
If I'm understanding your comment correctly, by "ARC references of various kinds" you mean `__strong`, `__weak`, `__unsafe_unretained`? That's what this function is applying; AFAICT the `VarDecl *prop` coming into this function is a synthesized decl with the bare unqualified type, and this function examines the qualifiers to determine which `ReferenceStorageType` it should be wrapped in.

Are you saying that `Impl.createDeclWithClangNode` might already be importing `ReferenceStorageType`s in some cases, before I reach here?

Reply:
Oh, I’m sorry, brain fart.